sysbuild: Fix issue with *_ROOT values not propagating

nordicjm opened the pull request:
Fixes an issue where variables like BOARD_ROOT would be provided to sysbuild but would then be lost on target images.

Fixes #53780

nordicjm committed:
sysbuild: Fix issue with *_ROOT values not propagating

Fixes an issue where variables like BOARD_ROOT would be provided
to sysbuild but would then be lost on target images.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand reviewed (commented):
thanks for this.

The direction looks fine, but I notice much of the code is identical with `zephyr_get(...)`, so I think it would be nice if code could be shared between the functions.

Perhaps as simple as having a `LIST` (or `MERGE` / `COMBINE`) flag to indicate that results from all scopes should be returned in the return var.

After all `zephyr_get(...)` will also return a list if the variable to look up is a list.
So both functions can return lists, the difference between them is that one will merge the results from all scope (local, cache, sysbuild) whereas the other will return just from a single scope (but may still return a list)

nordicjm committed:
doc: release: 3.4: Add note on build system changes.

Adds a note that *_ROOT variables are now supported properly on an
image basis when using sysbuild, and that sysbuild_cache_set has
been added, and that zephyr_get has had a MERGE option added.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2536,6 +2536,47 @@ function(zephyr_get variable)
+    endif()
+  endif()
+
+  get_property(from_sysbuild TARGET sysbuild_cache PROPERTY ${variable})
```
Moved inside if()

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2536,6 +2536,47 @@ function(zephyr_get variable)
+    list(APPEND ${variable} ${from_sysbuild})
+
+  if(DEFINED sysbuild_${variable})
+    list(APPEND ${variable} ${sysbuild_${variable}})
```
Added some comments, and yes. I have actually flipped the order to make the sysbuild setting be last when the MERGE option is used, so that priority of the image-specific configuration is higher

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2492,8 +2494,14 @@ endfunction()
 # using `-DZEPHYR_TOOLCHAIN_VARIANT=<val>`, then the value from the cache is
 # returned.
+#
+# There is an exclusion to the above rule when the MERGE option is used, in
+# which case the Sysbuild defined variable will be appended to the list last,
+# not first, this is so that local configuration to images will override the
+# default Sysbuild configuration, for example if a board directory is specified
+# on an image-level basis.
```
I disagree on this.

Firstly, it's not just the sysbuild variables that is appended last, also any cache variables.

This means that if a user does `cmake -DFOO=bar <sample>` and `sample` has `set(FOO baz)`, then order becomes `baz;bar`, and thus `baz` has highest precedence when using `zephyr_get(FOO MERGE)`.
That prevents users from enforcing a specific value from having highest precedence.

In such case, the value provided by the user (in this case `bar`) should take precedence over the sample defined value (`baz`).

A use-case could is the `BOARD_ROOT` where user can specify extra roots on command line.
Today, BOARD_ROOT specified on command line take highest precedence but if `BOARD_ROOT` starts using `zephyr_get(BOARD_ROOT MERGE)` only on sysbuild, then that would suddenly be a behavioral change in the build system. 

If a user today does:
`west build ... <sample> -- -DBOARD_ROOT=<my_custom_root>`
then that user should experience the same behavior when building that same sample using sysbuild, that is:
`west build ... --sysbuild <sample> -- -DBOARD_ROOT=<my_custom_root>`

With this PR, the order will be different between the two runs.
First run will have `<my_custom_root>` first in list, second run will have it last in list.

nordicjm replied:
I tested this and the order changing does not happen, using `cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -DAPP_DIR=.. -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -Dblob_BOARD_ROOT=/tmp/cc/example-application -DBOARD_ROOT=/tmp/bb/nrfxlib/ /tmp/aa/zephyr/share/sysbuild/` I correctly see `/tmp/bb/nrfxlib/` for sysbuild, `/tmp/bb/nrfxlib/;/tmp/aa/zephyr` for mcuboot, `/tmp/bb/nrfxlib/;/tmp/aa/zephyr` for the main app and `/tmp/cc/example-application;/tmp/bb/nrfxlib/;/tmp/aa/zephyr` for the blob target

nordicjm replied:
Have reworked this and made the order the same for both.

tejlmand commented on cmake/modules/root.cmake:
```diff
@@ -36,6 +36,14 @@ zephyr_file(APPLICATION_ROOT ARCH_ROOT)
 # Convert paths to absolute, relative from APPLICATION_SOURCE_DIR
 zephyr_file(APPLICATION_ROOT SCA_ROOT)
 
+if(SYSBUILD)
```
I don't believe fetching of roots should be depending on whether sysbuild is in use.

Why should a regular Zephyr CMake build not ensure same order of precedence regarding roots, as is done when sysbuild is in use ?

nordicjm replied:
Fixed

tejlmand reviewed (changes requested):
I think values defined by the user should have highest precedence, just like they have today wrt ROOTs defined.

Another issue I noticed is how `zephyr_get(<var> MERGE)` behaves compared to `zephyr_get(<var>)`.

If I do:
```
set(FOO bar)
zephyr_get(FOO)
print(FOO)
zephyr_get(FOO)
print(FOO)
```
I get:
```
-- FOO: bar
-- FOO: bar
```
but if I change to `zephyr_get(FOO MERGE)`, i'll get an ever growing list:
```
-- FOO: bar;bar
-- FOO: bar;bar;bar;bar
```

nordicjm committed:
sysbuild: Add sysbuild_cache_set CMake function

Adds a new CMake helper function that can be used from sysbuild to
set or update sysbuild CMake cache variables.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -380,5 +380,58 @@ function(sysbuild_module_call)
+    message(FATAL_ERROR "sysbuild_set(VAR <var> APPEND ...) missing required arguments to append")
+  elseif(VARS_LIST AND NOT VARS_APPEND)
+    message(FATAL_ERROR "sysbuild_set(VAR <var> APPEND LIST ...) is required for the LIST option")
+  elseif(NOT VARS_VAR)
+    message(FATAL_ERROR "sysbuild_set(VAR <var> ...) missing required VAR variable")
+  endif()
```
we now have helper macros in place to simplify testing and ensure some consistency in error messaging, so let's start to use them 😉 :
```suggestion
zephyr_check_arguments_required("sysbuild_set" VARS VAR)
```

nordicjm replied:
Fixed

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -380,5 +380,58 @@ function(sysbuild_module_call)
+    set(var_new ${VARS_UNPARSED_ARGUMENTS})
+  endif()
 
+  set(${VARS_VAR} ${var_new} CACHE ${var_type} "" FORCE)
```
If the variable already has a help text in the cache, then let's keep that help text by fetching the property value.
https://cmake.org/cmake/help/latest/prop_cache/HELPSTRING.html
```suggestion
  get_property(var_help CACHE ${VARS_VAR} PROPERTY HELPSTRING)
  set(${VARS_VAR} ${var_new} CACHE ${var_type} "${var_help}" FORCE)
```

nordicjm replied:
Fixed

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -380,5 +380,58 @@ function(sysbuild_module_call)
+  if(VARS_APPEND)
+    set(var_new $CACHE{${VARS_VAR}})
+    if(VARS_LIST)
+      list(APPEND var_new ${VARS_UNPARSED_ARGUMENTS})
```
I assume that this results in an ever expanding list in the CMake cache.
Maybe the value should only be appended if it doesn't exists in the cache list already.

Something like:
```suggestion
      foreach(new_val ${VARS_UNPARSED_ARGUMENTS})
        if(NOT new_val IN_LIST var_new)
          list(APPEND var_new ${new_val})
        endif()
      endforeach()
```
Note: code snippet above has not been validated.

nordicjm replied:
Have added an option to remove duplicates but also allow to keep them e.g. someone might be construct a command line which uses the same argument multiple times for different values

tejlmand replied:
>  e.g. someone might be construct a command line which uses the same argument multiple times for different values

makes a lot of sense.
Although when working with cache values and CMake re-runs, this is one of the tricky parts to get right, and we want to avoid ever growing list because of CMake re-runs.

CMake lists are effectively `;`-separated strings, so an option can be to search the list for pairs / n-length lists before appending, for example using: `string(FIND ...)`.

So you could search the list for such pairs, and if already existing, then not adding the duplicate.

As example:
```
set(list_var --foo)
set(bar1 --bar once)
set(bar2 --bar twice ${bar2})

list(APPEND list_var ${bar1})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            

list(APPEND list_var ${bar2})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            

list(APPEND list_var ${bar1})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            

list(APPEND list_var ${bar2})
list(REMOVE_DUPLICATES list_var)                                                                                                                                                            
```

gives the following list `--foo;--bar;once;twice`, and not `--foo;--bar;once;--bar;twice`

But if using `string(FIND ...)` before adding a new set of arguments, similar to this:
```
string(FIND "${list_var}" "${bar1}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar1})                                                                                                                                                               
endif()                                                                                                                                                                                       
                                                                                                                                                                                              
string(FIND "${list_var}" "${bar2}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar2})                                                                                                                                                               
endif()                                                                                                                                                                                       
                                                                                                                                                                                              
string(FIND "${list_var}" "${bar1}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar1})                                                                                                                                                               
endif()                       

string(FIND "${list_var}" "${bar2}" index)                                                                                                                                                    
if(${index} EQUAL -1)                                                                                                                                                                         
  list(APPEND list_var ${bar2})                                                                                                                                                               
endif()                                                                                                                                                                                       
```
then we avoid duplicated entries / ever growing list, while obtaining exactly the list we want, which is: `--foo;--bar;once;--bar;twice`

tejlmand replied:
Note: the `REMOVE_DUPLICATES` flag still makes a lot of sense to ensure a list of unique items.

The addition proposed in the above comment is to avoid ever growing lists of identical options / sub-lists / strings to be added on CMake re-runs.

nordicjm replied:
Have added this but I'm not particularly happy with how the functionality differs from list(APPEND ...) - not that I would envisage someone using the cmake cache for a command but someone might have a command added with verbosity levels which get increased, so "-v" might be added, then later on "-v" might get added, then finally another 2 "-v"'s get added, all done without removing duplicates, but with this code right now only the first "-v" would be added. But to counter that if you did a cmake rerun without the string check you would end up with "-v -v -v -v -v -v -v -v" on the rerun so is a challenge.

tejlmand replied:
>  but I'm not particularly happy with how the functionality differs from list(APPEND ...) 

no, but there is probably a good reason why CMake doesn't support `CACHE` and `FORCE` together with the [list](https://cmake.org/cmake/help/latest/command/list.html) function.

This is hinting that we are doing something we probably shouldn't be doing in the first place with this feature 😉 .
However, as we need the functionality provided with this PR, then I believe it's better to introduce `sysbuild_cache_set()` and limit the functionality so we prevent what would normally be unwanted behavior.

Those that need an ever growing list can create one such list by using CMake's regular `set()` function.
Remember, this PR is merely just offering an improved CMake cache handling for sysbuild with a built-in safety net.

> but with this code right now only the first "-v" would be added. 

Actually, if you do proper quoting of variables in `sysbuild_cache_set()`, then it should be possible to do:
```
sysbuild_cache_set(VAR CACHE_VAR APPEND "-v")
sysbuild_cache_set(VAR CACHE_VAR APPEND "-v\;")
```
or alternatively:
```
sysbuild_cache_set(VAR CACHE_VAR APPEND "-v;-v")
```

although in the last alternative user would need to ensure a single `-v` har not been added already.

tejlmand reviewed (changes requested):
Spotted some details in the implementation.

Regarding the `LOCAL` and `GLOBAL` flags, I think this should be updated in existing function description as I understand that current behavior is not precisely described.
I can create a dedicated PR with such improvement if you like.in

nordicjm commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -380,5 +380,58 @@ function(sysbuild_module_call)
+  cmake_parse_arguments(VARS "APPEND;LIST" "VAR" "" ${ARGN})
+
+  if(NOT VARS_UNPARSED_ARGUMENTS AND VARS_APPEND)
+    message(FATAL_ERROR "sysbuild_set(VAR <var> APPEND ...) missing required arguments to append")
```
The first if checks if the append option is set and disallows no argument but allows set with no argument so is equivalent to `set(ABC)`, the LIST APPEND disallows an empty parameter but is there really a need for an empty parameter in a list?

tejlmand replied:
>  but is there really a need for an empty parameter in a list?

you see all sorts of variants on how to set variables, and feed that into other function calls (both in / outside loops).
For example by conditionally set value on a var and then feed it to another call without a check, like here:
https://github.com/zephyrproject-rtos/zephyr/blob/82101f45e4996713c2b9f3b455a9ff593311ca5f/CMakeLists.txt#L950-L958
which is then used here regardless if defined or not:
https://github.com/zephyrproject-rtos/zephyr/blob/82101f45e4996713c2b9f3b455a9ff593311ca5f/CMakeLists.txt#L968
I know in this specific case there are also other values going into the final call.
It's more to show a variable use without knowing if it has actually been defined.

Or by having a return var populated by a function, for example:
```
special_function(IN ${input} OUT output)
sysbuild_cache_set(VAR CACHE_VAR APPEND LIST ${output})
```

in such a case it could be that `special_function()` returns nothing, some functions may not return anything under certain conditions.

If empty vars are not allowed users will have to do:
```
special_function(IN ${input} OUT output)
if(DEFINED output)
  sysbuild_cache_set(VAR CACHE_VAR APPEND LIST ${output})
endif()
```

Because of such use-cases, I see little reason to impose this restriction, especially because CMake's own `list(APPEND ...)` doesn't have such restriction, and I value consistency whenever possible.

nordicjm replied:
Have reworked it

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -380,5 +380,58 @@ function(sysbuild_module_call)
+endfunction()
+
+# Usage:
+#   sysbuild_cache_set(VAR <variable> [APPEND [LIST]] <value>)
```
> It is optional

my bad, I had read this:
https://github.com/zephyrproject-rtos/zephyr/blob/9ac601de052a2fd486c083c1995810d84255b51a/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L415
as `(VARS_APPEND AND NOT VARS_LIST)` which would have made both arguments required if APPEND was used.

> if LIST is not provided, is treated as a string e.g. the result would be "abc" + "eeq" = "abceeq" 

although the intention is good, i'm very concerned with such approach, because if any CMake code does:
```
sysbuild_cache_set(VAR CACHE_STRING APPEND "hello") 
```
you will get an ever growing string on each CMake re-run.
And you cannot string de-duplication safely, like you can do with list items.

nordicjm replied:
in that case I will remove appending strings

nordicjm replied:
Removed string appending

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2484,6 +2486,8 @@ endfunction()
 #     blinky image.
 #   If no sysbuild scope is specified, GLOBAL is assumed.
+#   If using MERGE then SYSBUILD BOTH can be used which will get both the local
+#   and global sysbuild scope variables (in that order).
```
Fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2493,39 +2497,69 @@ endfunction()
+    elseif(NOT DEFINED GET_VAR_MERGE AND
+           ${GET_VAR_SYSBUILD} STREQUAL "BOTH"
+    )
+      message(FATAL_ERROR "zephyr_get(... SYSBUILD) can only use BOTH with MERGE.")
```
Both option was removed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2493,39 +2497,69 @@ endfunction()
     if(NOT (${GET_VAR_SYSBUILD} STREQUAL "GLOBAL" OR
-            ${GET_VAR_SYSBUILD} STREQUAL "LOCAL")
+            ${GET_VAR_SYSBUILD} STREQUAL "LOCAL" OR
+            ${GET_VAR_SYSBUILD} STREQUAL "BOTH")
```
Fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2493,39 +2497,69 @@ endfunction()
 # returned.
 function(zephyr_get variable)
-  cmake_parse_arguments(GET_VAR "" "SYSBUILD" "" ${ARGN})
+  cmake_parse_arguments(GET_VAR "MERGE" "SYSBUILD" "" ${ARGN})
```
Variable is now cleared at start of function

tejlmand replied:
minor nit: you added the clearing of the var in the `sysbuild_cache_set()` commit here: https://github.com/zephyrproject-rtos/zephyr/pull/56801/commits/a0ec825aa62a5eee60769edb5fa312153e1c1585

so if you anyway need to do a PR update, then would be nice to have placed in first commit.

nordicjm replied:
Juggling too many changes, fixed.

tejlmand replied:
> Variable is now cleared at start of function

just discovered, only cleared, this part was missed: `with the introduction of MERGE we should probably read the existing local scope var`
(the sentence should probably had explicitly stated that the var should be added again on the return list)

If you try running `https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/application_development/out_of_tree_board` you'll notice that it picks up the board from ZEPHYR_BASE, and not the oot board, that is the following is printed:
```
-- Found BOARD.dts: /projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
```

instead of the expected:
```
-- Found BOARD.dts: /projects/github/ncs/zephyr/samples/application_development/out_of_tree_board/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
```

Note, the oot_board sample should probably be reworked to do a proper test, cause the test suffers two flaws:
- it uses a board name identical to an existing Zephyr board, meaning boards can be picked up from wrong location and that is not detected in the test.
- The application source dir is default added as a search root, meaning this line doesn't make much sense:
   https://github.com/zephyrproject-rtos/zephyr/blob/3bf7f83f213f96daa9a17cd3ddd4cff0a200bb9d/samples/application_development/out_of_tree_board/CMakeLists.txt#L6

I will take a look at this in relation to the snippets work.

tejlmand replied:
Ref: https://github.com/zephyrproject-rtos/zephyr/issues/57647

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -384,5 +384,70 @@ function(sysbuild_module_call)
+#                    within the list before saving to the cache.
+# <value>:           Value to set/update.
+function(sysbuild_cache_set)
+  cmake_parse_arguments(VARS "APPEND;REMOVE_DUPLICATES" "VAR" "" ${ARGN})
```
I believe quoting would be recommended here in if we think it is important to support the `"\;"` use-case. 

```suggestion
  cmake_parse_arguments(VARS "APPEND;REMOVE_DUPLICATES" "VAR" "" "${ARGN}")
```

nordicjm replied:
Looked through all the files in `cmake` in zephyr and none of them quote this parameter

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -384,5 +384,70 @@ function(sysbuild_module_call)
+  endif()
+
+  if(VARS_APPEND)
+    set(var_new $CACHE{${VARS_VAR}})
```
quoting is definitely needed here:
```suggestion
    set(var_new "$CACHE{${VARS_VAR}}")
```

else the content of the cache will be treated as arguments to the `set()` function, which is bad if the var value contains words like `cache`, `force`, and similar.

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -384,5 +384,70 @@ function(sysbuild_module_call)
+      return()
+    endif()
+
+    list(APPEND var_new ${VARS_UNPARSED_ARGUMENTS})
```
please check if quoting should be added here.

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -384,5 +384,70 @@ function(sysbuild_module_call)
+    set(var_new ${VARS_UNPARSED_ARGUMENTS})
+  endif()
 
+  set(${VARS_VAR} ${var_new} CACHE ${var_type} "${var_help}" FORCE)
```
I think quoting should be added here, but please verify.
```suggestion
  set(${VARS_VAR} "${var_new}" CACHE ${var_type} "${var_help}" FORCE)
```

nordicjm replied:
Added lots of quotes

tejlmand reviewed (changes requested):
Discovered that quoting is needed in various places when handling variables.

An unquoted variable is expanded and treated as multiple arguments to the functions, whereas a quoted variable will be treated as a single list argument.

For example the following code:
```
sysbuild_cache_set(VAR PR_TEST APPEND "a")
sysbuild_cache_set(VAR PR_TEST APPEND "CACHE;STRING;part of var")
sysbuild_cache_set(VAR PR_TEST APPEND "b")
```

results in the following cache value: `a;b` and not: `a;CACHE;STRING;part of var;b` as one would expect.

Please take time to do an extra verification of quoting, as it can both be that I've missed some locations or that some locations should still not be quoted.

tejlmand reviewed (approved):
🎉
